// ngx-infinite-scroller/esm5/src/app/scrolling-strategy/scrolling-to-top.js
import * as tslib_1 from "tslib";
import { filter } from 'rxjs/operators';
import { StrategyBase } from './strategy-base';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
var ScrollingToTop = /** @class */ (function (_super) {
    tslib_1.__extends(ScrollingToTop, _super);
    function ScrollingToTop(directive, state) {
        return _super.call(this, directive, state) || this;
    }
    ScrollingToTop.prototype.scrollDirectionChanged = function (scrollPairChanged) {
        var _this = this;
        return scrollPairChanged.pipe(filter(function (scrollPositions) {
            return _super.prototype.wasScrolledUp.call(_this, scrollPositions[0], scrollPositions[1]);
        }));
    };
    ScrollingToTop.prototype.scrollRequestZoneChanged = function (scrollDirectionChanged) {
        var _this = this;
        return scrollDirectionChanged.pipe(filter(function (scrollPositions) {
            return _super.prototype.isScrollUpEnough.call(_this, scrollPositions[1], _this.directive.scrollUpPercentilePositionTrigger);
        }));
    };
    ScrollingToTop.prototype.askForUpdate = function () {
        this.directive.onScrollUp.next();
    };
    ScrollingToTop.prototype.setInitialScrollPosition = function () {
        var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.BOTTOM);
        this.directive.scrollTo(initialScrollPositionValue);
    };
    ScrollingToTop.prototype.setPreviousScrollPosition = function () {
        var prevScrollPosition = this.state.previousScrollTop +
            (this.state.scrollHeight - this.state.previousScrollHeight);
        this.directive.scrollTo(prevScrollPosition);
    };
    return ScrollingToTop;
}(StrategyBase));
export { ScrollingToTop };
