// ngx-infinite-scroller/esm5/src/app/scrolling-strategy/scrolling-to-bottom.js
import * as tslib_1 from "tslib";
import { filter } from 'rxjs/operators';
import { StrategyBase } from './strategy-base';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
var ScrollingToBottom = /** @class */ (function (_super) {
    tslib_1.__extends(ScrollingToBottom, _super);
    function ScrollingToBottom(directive, state) {
        return _super.call(this, directive, state) || this;
    }
    ScrollingToBottom.prototype.scrollDirectionChanged = function (scrollPairChanged) {
        var _this = this;
        return scrollPairChanged.pipe(filter(function (scrollPositions) {
            return _super.prototype.wasScrolledDown.call(_this, scrollPositions[0], scrollPositions[1]);
        }));
    };
    ScrollingToBottom.prototype.scrollRequestZoneChanged = function (scrollDirectionChanged) {
        var _this = this;
        return scrollDirectionChanged.pipe(filter(function (scrollPositions) {
            return _super.prototype.isScrollDownEnough.call(_this, scrollPositions[1], _this.directive.scrollDownPercentilePositionTrigger);
        }));
    };
    ScrollingToBottom.prototype.askForUpdate = function () {
        this.directive.onScrollDown.next();
    };
    ScrollingToBottom.prototype.setInitialScrollPosition = function () {
        var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.TOP);
        this.directive.scrollTo(initialScrollPositionValue);
    };
    ScrollingToBottom.prototype.setPreviousScrollPosition = function () {
        var prevScrollPosition = this.state.previousScrollTop;
        this.directive.scrollTo(prevScrollPosition);
    };
    return ScrollingToBottom;
}(StrategyBase));
export { ScrollingToBottom };
