// ngx-infinite-scroller/esm5/src/app/scrolling-strategy/strategy-base.js
import isNumber from 'is-number';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
var StrategyBase = /** @class */ (function () {
    function StrategyBase(directive, state) {
        this.directive = directive;
        this.state = state;
    }
    StrategyBase.prototype.wasScrolledDown = function (prevPos, currentPos) {
        return prevPos.scrollTop < currentPos.scrollTop;
    };
    StrategyBase.prototype.wasScrolledUp = function (prevPos, currentPos) {
        return !this.wasScrolledDown(prevPos, currentPos);
    };
    StrategyBase.prototype.isScrollDownEnough = function (pos, scrollPositionTrigger) {
        return ((pos.scrollTop + pos.clientHeight) / pos.scrollHeight) > (scrollPositionTrigger / 100);
    };
    StrategyBase.prototype.isScrollUpEnough = function (pos, scrollPositionTrigger) {
        return (pos.scrollTop / pos.scrollHeight) < (scrollPositionTrigger / 100);
    };
    StrategyBase.prototype.getInitialScrollPositionValue = function (defaultScrollPosition) {
        var initialScrollPosition = this.directive.initialScrollPosition;
        if (isNumber(initialScrollPosition)) {
            return Number(initialScrollPosition);
        }
        var initialScrollPositions = this.getInitialScrollPositions();
        if (initialScrollPosition === InitialScrollPosition.DEFAULT) {
            return initialScrollPositions[defaultScrollPosition];
        }
        return initialScrollPositions[initialScrollPosition];
    };
    StrategyBase.prototype.getInitialScrollPositions = function () {
        var _a;
        var _b = this.state, scrollHeight = _b.scrollHeight, clientHeight = _b.clientHeight;
        return _a = {},
            _a[InitialScrollPosition.TOP] = 0,
            _a[InitialScrollPosition.MIDDLE] = scrollHeight / 2 - clientHeight / 2,
            _a[InitialScrollPosition.BOTTOM] = scrollHeight,
            _a;
    };
    return StrategyBase;
}());
export { StrategyBase };
