// ngx-infinite-scroller/esm5/src/app/scroll-height-listener/scroll-height-listener.js
var ScrollHeightListener = /** @class */ (function () {
    function ScrollHeightListener(directive, state) {
        this.directive = directive;
        this.state = state;
        this.DEFAULT_REQUEST_TIMEOUT = 30000;
    }
    ScrollHeightListener.prototype.start = function () {
        var _this = this;
        this.listener = window.requestAnimationFrame(this.listen.bind(this));
        if (!this.httpRequestTimeout) {
            this.httpRequestTimeout = setTimeout(function () {
                _this.stopIfRequestTimeout();
            }, this.DEFAULT_REQUEST_TIMEOUT);
        }
    };
    ScrollHeightListener.prototype.stop = function () {
        window.cancelAnimationFrame(this.listener);
        clearTimeout(this.httpRequestTimeout);
        this.httpRequestTimeout = null;
    };
    ScrollHeightListener.prototype.listen = function () {
        if (this.state.previousScrollHeight !== this.state.scrollHeight) {
            this.stop();
            this.directive.onScrollbarHeightChanged();
        }
        else {
            this.start();
        }
    };
    ScrollHeightListener.prototype.stopIfRequestTimeout = function () {
        if (!this.state.previousScrollPositionpUpdated) {
            this.stop();
        }
    };
    return ScrollHeightListener;
}());
export { ScrollHeightListener };
