// ngx-infinite-scroller/esm2015/src/app/scrolling-strategy/scrolling-to-top.js
import { filter } from 'rxjs/operators';
import { StrategyBase } from './strategy-base';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
export class ScrollingToTop extends StrategyBase {
    constructor(directive, state) {
        super(directive, state);
    }
    scrollDirectionChanged(scrollPairChanged) {
        return scrollPairChanged.pipe(filter((scrollPositions) => {
            return super.wasScrolledUp(scrollPositions[0], scrollPositions[1]);
        }));
    }
    scrollRequestZoneChanged(scrollDirectionChanged) {
        return scrollDirectionChanged.pipe(filter((scrollPositions) => {
            return super.isScrollUpEnough(scrollPositions[1], this.directive.scrollUpPercentilePositionTrigger);
        }));
    }
    askForUpdate() {
        this.directive.onScrollUp.next();
    }
    setInitialScrollPosition() {
        const initialScrollPositionValue = super.getInitialScrollPositionValue(InitialScrollPosition.BOTTOM);
        this.directive.scrollTo(initialScrollPositionValue);
    }
    setPreviousScrollPosition() {
        const prevScrollPosition = this.state.previousScrollTop +
            (this.state.scrollHeight - this.state.previousScrollHeight);
        this.directive.scrollTo(prevScrollPosition);
    }
}
