// ngx-infinite-scroller/esm2015/src/app/scrolling-strategy/strategy-base.js
import isNumber from 'is-number';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
export class StrategyBase {
    constructor(directive, state) {
        this.directive = directive;
        this.state = state;
    }
    wasScrolledDown(prevPos, currentPos) {
        return prevPos.scrollTop < currentPos.scrollTop;
    }
    wasScrolledUp(prevPos, currentPos) {
        return !this.wasScrolledDown(prevPos, currentPos);
    }
    isScrollDownEnough(pos, scrollPositionTrigger) {
        return ((pos.scrollTop + pos.clientHeight) / pos.scrollHeight) > (scrollPositionTrigger / 100);
    }
    isScrollUpEnough(pos, scrollPositionTrigger) {
        return (pos.scrollTop / pos.scrollHeight) < (scrollPositionTrigger / 100);
    }
    getInitialScrollPositionValue(defaultScrollPosition) {
        const { initialScrollPosition } = this.directive;
        if (isNumber(initialScrollPosition)) {
            return Number(initialScrollPosition);
        }
        const initialScrollPositions = this.getInitialScrollPositions();
        if (initialScrollPosition === InitialScrollPosition.DEFAULT) {
            return initialScrollPositions[defaultScrollPosition];
        }
        return initialScrollPositions[initialScrollPosition];
    }
    getInitialScrollPositions() {
        const { scrollHeight, clientHeight } = this.state;
        return {
            [InitialScrollPosition.TOP]: 0,
            [InitialScrollPosition.MIDDLE]: scrollHeight / 2 - clientHeight / 2,
            [InitialScrollPosition.BOTTOM]: scrollHeight,
        };
    }
}
